File: src/main/java/me/melijn/jda/utils/TaskManager.java
package me.melijn.jda.utils;

import com.google.common.util.concurrent.ThreadFactoryBuilder;

import java.time.LocalDate;
import java.util.concurrent.*;
import java.util.function.Function;
import java.util.function.Predicate;

public class TaskManager {

    private final MessageHelper messageHelper;
    private final Function<String, ThreadFactory> threadFactory = name -> new ThreadFactoryBuilder().setNameFormat("[" + name + "-Pool-%d] ").build();
    private final ExecutorService executorService = Executors.newCachedThreadPool(threadFactory.apply("Task"));
    private final ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(10, threadFactory.apply("Rep"));

    public TaskManager(MessageHelper messageHelper) {
        this.messageHelper = messageHelper;
    }

    public void scheduleRepeating(final Runnable runnable, final long period) {
        scheduledExecutorService.scheduleAtFixedRate(new Task(messageHelper, runnable), 0, period, TimeUnit.MILLISECONDS);
    }

    public void scheduleRepeating(final Runnable runnable, final long initialDelay, final long period) {
        scheduledExecutorService.scheduleAtFixedRate(new Task(messageHelper, runnable), initialDelay, period, TimeUnit.MILLISECONDS);
    }

    public void async(final Runnable runnable) {
        executorService.submit(new Task(messageHelper, runnable));
    }

    public void async(final Runnable runnable, final long after) {
        scheduledExecutorService.schedule(new Task(messageHelper, runnable), after, TimeUnit.MILLISECONDS);
    }

    public ExecutorService getExecutorService() {
        return executorService;
    }

    public ScheduledExecutorService getScheduledExecutorService() {
        return scheduledExecutorService;
    }
}

File: src/main/java/me/melijn/melijnbot/objects/threading/TaskManager.kt
package me.melijn.melijnbot.objects.threading

import com.google.common.util.concurrent.ThreadFactoryBuilder
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit


class TaskManager {

    private val threadFactory = { name: String -> ThreadFactoryBuilder().setNameFormat("[$name-Pool-%d] ").build() }
    private val executorService = Executors.newCachedThreadPool(threadFactory.invoke("Task"))
    private val scheduledExecutorService = Executors.newScheduledThreadPool(15, threadFactory.invoke("Repeater"))

    fun scheduleRepeating(runnable: Runnable, periodMillis: Long) {
        scheduledExecutorService.scheduleAtFixedRate(Task(runnable), 0, periodMillis, TimeUnit.MILLISECONDS)
    }

    fun scheduleRepeating(runnable: Runnable, afterMillis: Long, periodMillis: Long) {
        scheduledExecutorService.scheduleAtFixedRate(Task(runnable), afterMillis, periodMillis, TimeUnit.MILLISECONDS)
    }

    fun async(func: () -> Unit) {
        executorService.submit(Task(Runnable(func)))
    }

    fun asyncAfter(afterMillis: Long, func: () -> Unit) {
        scheduledExecutorService.schedule(Task(Runnable(func)), afterMillis, TimeUnit.MILLISECONDS)
    }

    fun getExecutorService(): ExecutorService {
        return executorService
    }

    fun getScheduledExecutorService(): ScheduledExecutorService {
        return scheduledExecutorService
    }
}